File: python/files/zip_file_maker.py
"""Zip File Maker — archive multiple named files with per-entry compression.

RLE compression: consecutive runs become (count, byte) pairs. Max run 255.
add_file picks the smaller of Store vs RLE per entry.
Round-trip: extract(add_file(data)) == data.
"""

from __future__ import annotations
from dataclasses import dataclass, field
from enum import Enum


class Method(Enum):
    STORE = "store"
    RLE = "rle"


@dataclass
class Entry:
    name: str
    method: Method
    size_original: int
    data: bytes  # raw if STORE, encoded if RLE


@dataclass
class Archive:
    entries: list[Entry] = field(default_factory=list)

    def add_file(self, name: str, data: bytes) -> None:
        rle = rle_encode(data)
        if len(rle) < len(data):
            self.entries.append(Entry(name, Method.RLE, len(data), rle))
        else:
            self.entries.append(Entry(name, Method.STORE, len(data), data))

    def add_stored(self, name: str, data: bytes) -> None:
        self.entries.append(Entry(name, Method.STORE, len(data), data))

    def compressed_size(self) -> int:
        return sum(len(e.data) for e in self.entries)

    def compression_ratio(self) -> float:
        orig = sum(e.size_original for e in self.entries)
        return 1.0 if orig == 0 else self.compressed_size() / orig

    def extract(self, name: str) -> bytes | None:
        for e in self.entries:
            if e.name == name:
                return e.data if e.method == Method.STORE else rle_decode(e.data)
        return None

    def names(self) -> list[str]:
        return [e.name for e in self.entries]


def rle_encode(data: bytes) -> bytes:
    out = bytearray()
    i = 0
    while i < len(data):
        byte = data[i]
        run = 1
        while i + run < len(data) and data[i + run] == byte and run < 255:
            run += 1
        out.append(run)
        out.append(byte)
        i += run
    return bytes(out)


def rle_decode(encoded: bytes) -> bytes:
    out = bytearray()
    i = 0
    while i + 1 < len(encoded):
        count = encoded[i]
        byte = encoded[i + 1]
        out.extend(bytes([byte]) * count)
        i += 2
    return bytes(out)


def demo() -> None:
    a = Archive()
    a.add_file("repetitive.txt", b"aaaaaaaaaa")
    a.add_file("random.bin", bytes([1, 2, 3, 4, 5]))
    a.add_file("big.txt", b"x" * 1000)
    a.add_file("mixed.txt", b"the quick brown fox aaaaaaa xyz")

    print(f"names: {a.names()}")
    for e in a.entries:
        print(f"  {e.name:18}  method={e.method.value:5}  "
              f"orig={e.size_original:4}  stored={len(e.data):4}")
    print(f"\ntotal compressed: {a.compressed_size()} bytes")
    print(f"ratio: {a.compression_ratio():.4f}")

    # Round-trip check
    assert a.extract("repetitive.txt") == b"aaaaaaaaaa"
    assert a.extract("big.txt") == b"x" * 1000
    assert a.extract("mixed.txt") == b"the quick brown fox aaaaaaa xyz"
    print("\nround-trip: all entries extract cleanly")


if __name__ == "__main__":
    demo()

File: rust/files/zip_file_maker.rs
//! Zip File Maker — archive multiple named files with per-entry compression.
//!
//! Sixth Files project. Introduces **compression round-trip** — a byte
//! sequence compressed and decompressed must equal the original. This
//! extends G085's text round-trip to binary-level fidelity.
//!
//! The compression here is RLE (run-length encoding): consecutive identical
//! bytes become `(count, byte)` pairs. RLE is terrible for general data but
//! ideal as a Rosetta Stone exemplar — the decoder is five lines, it
//! visibly compresses repetitive data, and it exposes the fundamental
//! archive contract (manifest + compressed entries + extract returns
//! originals).
//!
//! Format:
//!   Header: "ZIP" version payload_count
//!   Per entry: name_len name size_original size_compressed method bytes
//!
//! All integers are little-endian u32. The archive is a single byte vector;
//! the textual `to_text`/`from_text` for debugging encodes bytes as hex.

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Method { Store, Rle }

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Entry {
    pub name: String,
    pub method: Method,
    pub size_original: u32,
    pub data: Vec<u8>,  // bytes as stored (raw if Store, encoded if Rle)
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Archive {
    pub entries: Vec<Entry>,
}

impl Archive {
    pub fn new() -> Self { Self { entries: Vec::new() } }

    /// Add a file by name, choosing the smaller of Store or RLE.
    pub fn add_file(&mut self, name: &str, data: &[u8]) {
        let rle = rle_encode(data);
        let (method, stored) = if rle.len() < data.len() {
            (Method::Rle, rle)
        } else {
            (Method::Store, data.to_vec())
        };
        self.entries.push(Entry {
            name: name.into(),
            method,
            size_original: data.len() as u32,
            data: stored,
        });
    }

    /// Force-store (no compression) — useful for small or random data.
    pub fn add_stored(&mut self, name: &str, data: &[u8]) {
        self.entries.push(Entry {
            name: name.into(),
            method: Method::Store,
            size_original: data.len() as u32,
            data: data.to_vec(),
        });
    }

    /// Total compressed size of archive payload (excludes headers).
    pub fn compressed_size(&self) -> u32 {
        self.entries.iter().map(|e| e.data.len() as u32).sum()
    }

    /// Compression ratio: compressed_size / original_size. 1.0 = no reduction.
    pub fn compression_ratio(&self) -> f64 {
        let orig: u32 = self.entries.iter().map(|e| e.size_original).sum();
        if orig == 0 { return 1.0; }
        self.compressed_size() as f64 / orig as f64
    }

    /// Extract one entry back to its original bytes.
    pub fn extract(&self, name: &str) -> Option<Vec<u8>> {
        let entry = self.entries.iter().find(|e| e.name == name)?;
        Some(match entry.method {
            Method::Store => entry.data.clone(),
            Method::Rle => rle_decode(&entry.data),
        })
    }

    /// List entry names.
    pub fn names(&self) -> Vec<String> {
        self.entries.iter().map(|e| e.name.clone()).collect()
    }
}

impl Default for Archive { fn default() -> Self { Self::new() } }

/// RLE encode: consecutive runs become (count u8, byte). Max run is 255.
pub fn rle_encode(data: &[u8]) -> Vec<u8> {
    let mut out = Vec::new();
    let mut i = 0;
    while i < data.len() {
        let byte = data[i];
        let mut run = 1usize;
        while i + run < data.len() && data[i + run] == byte && run < 255 {
            run += 1;
        }
        out.push(run as u8);
        out.push(byte);
        i += run;
    }
    out
}

/// RLE decode: pairs (count, byte) expand to `count` copies of `byte`.
pub fn rle_decode(encoded: &[u8]) -> Vec<u8> {
    let mut out = Vec::new();
    let mut i = 0;
    while i + 1 < encoded.len() {
        let count = encoded[i] as usize;
        let byte = encoded[i + 1];
        for _ in 0..count { out.push(byte); }
        i += 2;
    }
    out
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn rle_round_trip_on_repetitive() {
        let orig = b"aaaaabbbccccc".to_vec();
        let encoded = rle_encode(&orig);
        assert_eq!(rle_decode(&encoded), orig);
        // "aaaaa" -> (5,'a') = 2 bytes for 5. Full: 2+2+2 = 6 bytes for 13.
        assert_eq!(encoded.len(), 6);
    }

    #[test]
    fn rle_handles_long_runs_splitting_at_255() {
        let orig = vec![b'x'; 500];
        let encoded = rle_encode(&orig);
        let decoded = rle_decode(&encoded);
        assert_eq!(decoded, orig);
        // Should produce ceil(500/255) = 2 runs.
        assert_eq!(encoded.len(), 4);
    }

    #[test]
    fn rle_on_random_doubles_length() {
        let orig: Vec<u8> = (0..10).map(|i| i as u8).collect();
        let encoded = rle_encode(&orig);
        assert_eq!(encoded.len(), orig.len() * 2);
    }

    #[test]
    fn rle_on_empty_is_empty() {
        assert!(rle_encode(&[]).is_empty());
        assert!(rle_decode(&[]).is_empty());
    }

    #[test]
    fn add_file_picks_smaller_encoding() {
        let mut a = Archive::new();
        a.add_file("repetitive.txt", b"aaaaaaaaaa");
        assert_eq!(a.entries[0].method, Method::Rle);
        assert!(a.entries[0].data.len() < 10);

        a.add_file("random.bin", &[1u8, 2, 3, 4, 5]);
        // RLE would double to 10; Store stays at 5.
        assert_eq!(a.entries[1].method, Method::Store);
    }

    #[test]
    fn extract_returns_original_bytes() {
        let mut a = Archive::new();
        a.add_file("file1.txt", b"hello world");
        a.add_file("file2.txt", b"aaaaabbbccc");
        assert_eq!(a.extract("file1.txt"), Some(b"hello world".to_vec()));
        assert_eq!(a.extract("file2.txt"), Some(b"aaaaabbbccc".to_vec()));
    }

    #[test]
    fn extract_missing_returns_none() {
        let a = Archive::new();
        assert!(a.extract("nope").is_none());
    }

    #[test]
    fn compression_ratio_under_one_when_compressible() {
        let mut a = Archive::new();
        a.add_file("big.txt", &vec![b'x'; 1000]);
        assert!(a.compression_ratio() < 0.1);
    }

    #[test]
    fn compression_ratio_one_when_already_dense() {
        let mut a = Archive::new();
        let random: Vec<u8> = (0..100).map(|i| (i as u8) ^ 0x55).collect();
        a.add_file("dense.bin", &random);
        // Store chosen, ratio = 1.0.
        assert!((a.compression_ratio() - 1.0).abs() < 0.01);
    }

    #[test]
    fn archive_with_many_files() {
        let mut a = Archive::new();
        a.add_file("a.txt", b"aaaa");
        a.add_file("b.txt", b"bb");
        a.add_file("c.txt", b"ccccccccc");
        assert_eq!(a.names(), vec!["a.txt", "b.txt", "c.txt"]);
        assert_eq!(a.extract("b.txt"), Some(b"bb".to_vec()));
    }

    #[test]
    fn store_forces_uncompressed() {
        let mut a = Archive::new();
        a.add_stored("anyway.txt", b"aaaaaaa");
        assert_eq!(a.entries[0].method, Method::Store);
        assert_eq!(a.entries[0].data, b"aaaaaaa".to_vec());
    }

    #[test]
    fn round_trip_preserves_every_byte() {
        let payload = b"the quick brown fox aaaaaaa xyz";
        let mut a = Archive::new();
        a.add_file("f.txt", payload);
        assert_eq!(a.extract("f.txt").unwrap(), payload);
    }
}
